slotsmanager.c

```diff
@@ -434,7 +497,10 @@ slots_mark_garbage(int i)
     P[i].IsGarbage = 1;
     if(SLOTS_ENABLED(P[i].Type)) {
         BASESLOT(i)->IsGarbage = 1;
+        atomic_add_and_fetch(&SlotsManager->info[P[i].Type].garbage, 1);
     }
+    /*Increment the counter*/
+    atomic_add_and_fetch(&SlotsManager->garbage, 1);
```

Comment:
Why count for garbage here?

It's usually more efficient to avoid atomic operation and count the number of garbages if we have to run though the full particle storage anyways during GC.

If all we need is whether there is any garbage, then we don't need an atomic write -- just set has_garbage to 1.

Reply:
In either case you do as many as +1 operations -- the atomic version is hell a lot worse than the non-atomic version.

Reply:
The per-type atomics are there because we need to know if more than 10% of the slot is garbage to skip doing a gc.  The only other way to know that is to loop over the BASESLOT array. If we need to do that even if we are skipping the slot (which in practice we always are), gc becomes 50% slower and that completely dominates over the extra locking cost.

mark_slot_garbage is not called very often except during exchange, and the loop there is not threaded. Also when it is called there is another atomic access just before in fork, which will serialise.

 I can pretty easily avoid the overall garbage count though, if you think that is worth it. We could also just comment out the slots_gc_slots() - it is never called in practice because every PM step domain_decompose_full takes care of the garbage. What do you think?

Reply:
Indeed during tree walk, not often. Though I recall during exchange we might call this quite often -- since we mark all outgoing particles as garbage there. Though I am not sure how bad a performance hit it really will be since the exchange is likely a lot more expensive anyways. So I agree with you now on keeping the atomic locks.

Avoid having an overall garbage count reduces the amount of book keeping between GC cycles -- which sounds like a cleaner design -- remember the relief when we get rid of the book keeping code for TotNumPart and TimBinCount? However it appears that we need some of these to keep track if it is necessary to do a GC, so let's keep them. Could you change the name of variable to ngarbage to indicate it is the number of garbage and add a comment in the struct to say it is maintained by using mark() during cycles?

I am a bit supprised slots_gc_slots is no longer used. I need to digest this.

Reply:
The exchange algorithm uses gc to purge the out-going particles before receiving. So you can't really remove the routine -- otherwise we will run out of the slots quickly -- in cases where multiple iterations of exchange is required. 

Is the difference betweens slots_gc and slots_gc_sorted only that the latter sorts the particles before GC? If so perhaps it is cleaner to have a slots_sort_base(), and call both functions rather than having a larger function that does both.

Reply:
We do call it often during exchange - but that loop is single-threaded, so there is zero contention.

slots_gc_sorted() does a sort of the particles, moving garbage to the end. It sorts the base array and then it sorts the slots array. It doesn't use the O(n) garbage algorithm in slots_gc() at all, because it reorders rather than just compacting. I think I will add a comment to this effect.

So slots_gc_slots() is no longer called often in normal circumstances because to be called we need > 10% of the particles exchanged between two PM timesteps, and the displacement constraint  means that this doesn't happen often. 

You make a good point though that under low memory conditions with multiple exchange iterations we do need it - the slots could use up all the memory otherwise. In fact, I will make it explicit that it always runs when multiple iterations happen, even if it would compact by < 10% of the slots.